tdrs-backend/tdpservice/search_indexes/parsers/preparser.py

```diff
@@ -0,0 +1,262 @@
+        'blank':        {'type': 'string', 'required': True, 'regex': '^[ ]{9}$'},
+    }
+
+    validator = Validator(trailer_schema)
+
+    trailer = {
+        'title':        row[0:7],
+        'record_count': row[7:14],
+        'blank':        row[14:23],
+    }
+
+    is_valid = validator.validate(trailer)
+
+    # logger.debug("Trailer title: '%s'", trailer['title'])
+    # logger.debug("Trailer record count: '%s'", trailer['record_count'])
+    # logger.debug("Trailer blank: '%s'", trailer['blank'])
+    # logger.debug("Trailer errors: '%s'", validator.errors)
+
+    return is_valid, validator
+
+def get_header_row(datafile):
+    """Alters header row into string."""
+    # intentionally only reading first line of file
+    datafile.seek(0)
+    row = datafile.readline()
+    datafile.seek(0)  # reset file pointer to beginning of file
+
+    # datafile when passed via redis/db is a FileField which returns bytes
+    if isinstance(row, bytes):
+        row = row.decode()
+
+    if get_record_type(row) != 'HE':
+        raise ValueError('First line in file is not recognized as a valid header.')
+    elif len(row) != 24:
+        logger.debug("row: '%s' len: %d", row, len(row))
+        return False, {'preparsing': 'Header length incorrect.'}
+    row = row.strip('\n')
+
+    return True, row
+
+def get_trailer_row(datafile):
+    """Alters the trailer row into usable string."""
+    # certify/transform input row to be correct form/type
+
+    # Don't want to read whole file, just last line, only possible with binary
+    # Credit: https://openwritings.net/pg/python/python-read-last-line-file
+    # https://stackoverflow.com/questions/46258499/
+    try:
+        datafile.seek(-2, os.SEEK_END)  # Jump to the second last byte.
+        while datafile.read(1) != b'\n':  # Check if new line.
+            datafile.seek(-2, os.SEEK_CUR)   # Jump two bytes back
+    except OSError:  # Either file is empty or contains one line.
+        datafile.seek(0)
+        return False, {'preparsing': 'File too short or missing trailer.'}
+
+    # Having set the file pointer to the last line, read it in.
+    row = datafile.readline().decode()
+    datafile.seek(0)  # Reset file pointer to beginning of file.
+
+    logger.info("Trailer row: '%s'", row)
+    if get_record_type(row) != 'TR':
+        raise ValueError('Last row is not recognized as a trailer row.')
+    elif len(row) != 24:
+        logger.debug("row: '%s' len: %d", row, len(row))
+        return False, {'preparsing': 'Trailer length incorrect.'}
+    row = row.strip('\n')
+
+    return True, row
+
+def preparse(data_file, data_type, section):
+    """Validate metadata then dispatches file to appropriate parser."""
+    if isinstance(data_file, DataFile):
+        logger.debug("Beginning preparsing on '%s'", data_file.file.name)
+        datafile = data_file.file  # do I need to open() this?
+    elif isinstance(data_file, BufferedReader):
+        datafile = data_file
```

Comment:
The fact that data_file could be multiple data types and that it changes what type it is in this function is a bit of a red flag. This could be confusing to debug in the future. This might be a place you consider using types.

Reply:
The tests could submit a data file directly to the view like these ones we wrote for the email functionality

https://github.com/raft-tech/TANF-app/blob/4057f5d364e56e223c82bf4bc65d9f3f42a22df3/tdrs-backend/tdpservice/data_files/test/test_api.py#L239-L256

that expands the scope of the test but might help with getting different types for the file

Reply:
Tried to get the above post to work, no luck, pytest skips this class entirely. Open to pair on it @jtimpe

Reply:
really like to see why it that class is ignored